app/services/fact_importer.rb

```diff
@@ -40,26 +40,62 @@ def initialize(host, facts = {})
   # expect a facts hash
   def import!
     ActiveRecord::Base.transaction do
+      ensure_fact_names
       delete_removed_facts
       update_facts
       add_new_facts
     end
 
-    if @error
-      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
-      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s"), @host.name)
-    end
+    report_error(@error) if @error
     logger.info("Import facts for '#{host}' completed. Added: #{counters[:added]}, Updated: #{counters[:updated]}, Deleted #{counters[:deleted]} facts")
   end
 
+  def report_error(error)
+    Foreman::Logging.exception("Error during fact import for #{@host.name}", error)
+    raise ::Foreman::WrappedException.new(error, N_("Import of facts failed for host %s"), @host.name)
+  end
+
   # to be defined in children
   def fact_name_class
     raise NotImplementedError
   end
 
   private
 
-  attr_reader :host, :facts
+  attr_reader :host, :facts, :fact_names
+
+  def ensure_fact_names
+    initialize_fact_names
+
+    missing_keys = facts.keys - fact_names.keys
+    add_missing_fact_names(missing_keys)
+  end
+
+  def add_missing_fact_names(missing_keys)
+    missing_keys.sort.each do |fact_name|
```

Comment:
It's for `structured_fact_importer` - I want to make sure `a` is created before `a::b`.

Reply:
Then extract it into method which you override in structured fact, so plain does not sort? Just an optimization idea.

Reply:
I see now, still a quite expensive step for this request. Is it possible to avoid this? Let's say you won't construct `missing_keys` and iterate through `facts.keys` directly making sure that parents gets created first. Just an idea, this would also save up to N allocations and then sort.